financial_calculations/forward_curves.py

```diff
@@ -73,3 +74,104 @@ def objective(rate: float):
 
     return spot_rate_dates, spot_rates
 
+def bootstrap_finer_forward_curve(cmt_data, market_close_date, par_value, frequency='monthly', initial_guess=0.04, penalty=1.0):
+    """
+    Bootstraps a finer forward curve using a specified grid frequency (monthly/weekly).
+    This method penalizes large rate jumps to ensure smoother transitions between rates.
+
+    Parameters:
+    -----------
+    cmt_data : list of tuples
+        A list of (maturity_years, coupon_rate) tuples representing bond data.
+        e.g. [(1, 0.03), (2, 0.04), (3, 0.05)] for 1, 2, 3 year bonds with 3%, 4%, 5% coupons.
+    market_close_date : datetime or np.datetime64
+        The market close date for the bonds. If the date is a numpy datetime64[D], it's converted to a datetime object.
+    par_value : float
+        The par value of the bonds.
+    frequency : str, optional
+        The frequency of the date grid for the forward curve. Choices are 'monthly' (default) or 'weekly'.
+    initial_guess : float, optional
+        The initial guess for the spot rate in the optimization routine. Default is 0.04 (4%).
+    penalty : float, optional
+        A penalty parameter that penalizes large jumps in the discount rates to ensure smoother rate transitions.
+    
+    Returns:
+    --------
+    tuple:
+        A tuple containing two elements:
+        - spot_rate_dates (np.ndarray): Array of finer spot rate dates.
+        - spot_rates (np.ndarray): Array of spot rates corresponding to each spot rate date.
+
+    Raises:
+    -------
+    ValueError:
+        If the frequency is not 'monthly' or 'weekly'
+        If the optimization process fails to converge for a specific bond.
+    """
+
+    # If market_close_date is numpy datetime64, convert to datetime for use with relativedelta
+    if isinstance(market_close_date, np.datetime64): 
+        market_close_date_dt = market_close_date.astype(datetime)
+        market_close_date = datetime.combine(market_close_date_dt, datetime.min.time())  # Ensure time is set to 00:00:00
+
+    # Create a finer grid of dates (monthly/weekly) up to the longest bond maturity
```

Comment:
I get the feeling that there is too much code here.  the only difference between this code and coarse version is disc dates and the objective function.  everything else should be the same IIRC.